base/linalg/triangular.jl

```diff
@@ -149,65 +149,77 @@ istril(A::UnitLowerTriangular) = true
 
-function tril(A::UpperTriangular,k::Integer=0)
+function tril!(A::UpperTriangular,k::Integer=0)
     n = size(A,1)
     if abs(k) > n
         throw(ArgumentError("requested diagonal, $k, out of bounds in matrix of size ($n,$n)"))
     elseif k < 0
-        return UpperTriangular(zeros(A.data))
+        fill!(A.data,0)
+        return A
     elseif k == 0
         return UpperTriangular(diagm(diag(A)))
     else
-        return UpperTriangular(triu(tril(A.data,k)))
+        return UpperTriangular(triu!(tril!(A.data,k)))
     end
 end
+triu!(A::UpperTriangular,k::Integer=0) = UpperTriangular(triu!(triu!(A.data),k))
 
-triu(A::UpperTriangular,k::Integer=0) = UpperTriangular(triu(triu(A.data),k))
-
-function tril(A::UnitUpperTriangular,k::Integer=0)
+function tril!(A::UnitUpperTriangular,k::Integer=0)
     n = size(A,1)
     if abs(k) > n
         throw(ArgumentError("requested diagonal, $k, out of bounds in matrix of size ($n,$n)"))
     elseif k < 0
-        return UnitUpperTriangular(zeros(A.data))
+        fill!(A.data,0)
+        return UpperTriangular(A.data)
     elseif k == 0
         return UnitUpperTriangular(eye(A))
     else
-        return UnitUpperTriangular(triu(tril(A.data,k)))
+        return UnitUpperTriangular(triu!(tril!(A.data,k)))
     end
 end
 
-triu(A::UnitUpperTriangular,k::Integer=0) = UnitUpperTriangular(triu(triu(A.data),k))
+triu!(A::UnitUpperTriangular,k::Integer=0) = UnitUpperTriangular(triu!(triu!(A.data),k))
 
-function triu(A::LowerTriangular,k::Integer=0)
+function triu!(A::LowerTriangular,k::Integer=0)
     n = size(A,1)
     if abs(k) > n
         throw(ArgumentError("requested diagonal, $k, out of bounds in matrix of size ($n,$n)"))
     elseif k > 0
-        return LowerTriangular(zeros(A.data))
+        fill!(A.data,0)
+        return A
     elseif k == 0
         return LowerTriangular(diagm(diag(A)))
     else
-        return LowerTriangular(tril(triu(A.data,k)))
+        return LowerTriangular(tril!(triu!(A.data,k)))
     end
 end
 
-tril(A::LowerTriangular,k::Integer=0) = LowerTriangular(tril(tril(A.data),k))
+tril!(A::LowerTriangular,k::Integer=0) = LowerTriangular(tril!(tril!(A.data),k))
 
-function triu(A::UnitLowerTriangular,k::Integer=0)
+function triu!(A::UnitLowerTriangular,k::Integer=0)
     n = size(A,1)
     if abs(k) > n
         throw(ArgumentError("requested diagonal, $k, out of bounds in matrix of size ($n,$n)"))
     elseif k > 0
-        return UnitLowerTriangular(zeros(A.data))
+        fill!(A.data,0)
+        return LowerTriangular(A.data)
     elseif k == 0
         return UnitLowerTriangular(eye(A))
     else
-        return UnitLowerTriangular(tril(triu(A.data,k)))
+        return UnitLowerTriangular(tril!(triu!(A.data,k)))
     end
 end
 
-tril(A::UnitLowerTriangular,k::Integer=0) = UnitLowerTriangular(tril(tril(A.data),k))
+tril!(A::UnitLowerTriangular,k::Integer=0) = UnitLowerTriangular(tril!(tril!(A.data),k))
+
+tril(A::UpperTriangular,k::Integer=0)     = tril!(copy(A),k)
```

Comment:
should the generic fallbacks here https://github.com/JuliaLang/julia/blob/24a92a9f5df149b5a58ad1c4758de87599f5a9ac/base/linalg/generic.jl#L41-L44
be tweaked to cover all of these?